specification/app/resource-manager/Microsoft.App/preview/2023-05-02-preview/Patches.json

```diff
@@ -0,0 +1,392 @@
+        "parameters": [
+          {
+            "$ref": "../../../../../common-types/resource-management/v3/types.json#/parameters/ApiVersionParameter"
+          },
+          {
+            "$ref": "../../../../../common-types/resource-management/v3/types.json#/parameters/SubscriptionIdParameter"
+          },
+          {
+            "$ref": "../../../../../common-types/resource-management/v3/types.json#/parameters/ResourceGroupNameParameter"
+          },
+          {
+            "$ref": "./Builders.json#/parameters/BuilderNameParameter"
+          },
+          {
+            "$ref": "#/parameters/PatchNameParameter"
+          },
+          {
+            "$ref": "#/definitions/PatchConfiguration"
+          }
+        ],
+        "responses": {
+          "200": {
+            "description": "Patching operation completed successfully.",
+            "schema": {
+              "$ref": "#/definitions/PatchResource"
+            }
+          },
+          "201": {
+            "description": "Patching operation has been started.",
+            "schema": {
+              "$ref": "#/definitions/PatchResource"
+            },
+            "headers": {
+              "Azure-AsyncOperation": {
+                "type": "string"
+              }
+            }
+          },
+          "default": {
+            "description": "Common error response.",
+            "schema": {
+              "$ref": "./CommonDefinitions.json#/definitions/DefaultErrorResponse"
+            }
+          }
+        },
+        "x-ms-long-running-operation-options": {
+          "final-state-via": "azure-async-operation"
+        },
+        "x-ms-examples": {
+          "Patches_Apply_0": {
+            "$ref": "./examples/Patches_Apply.json"
+          }
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.App/builders/{builderName}/patches/{patchName}/delete": {
+      "put": {
+        "tags": [
+          "Patches"
+        ],
+        "operationId": "Patches_Delete",
+        "description": "Delete specific patch by patch id.",
+        "parameters": [
+          {
+            "$ref": "../../../../../common-types/resource-management/v3/types.json#/parameters/ApiVersionParameter"
+          },
+          {
+            "$ref": "../../../../../common-types/resource-management/v3/types.json#/parameters/SubscriptionIdParameter"
+          },
+          {
+            "$ref": "../../../../../common-types/resource-management/v3/types.json#/parameters/ResourceGroupNameParameter"
+          },
+          {
+            "$ref": "./Builders.json#/parameters/BuilderNameParameter"
+          },
+          {
+            "$ref": "#/parameters/PatchNameParameter"
+          },
+          {
+            "$ref": "#/definitions/PatchConfiguration"
+          }
+        ],
+        "responses": {
+          "200": {
+            "description": "Patch deletion completed successfully."
+          },
```

Comment:
nit: is deleting a patch not a long-running operation? I think I might need more information on how this is going to differ from deleting a build proxy resource (which @pauld-msft may be able to provide more context for), but I want to make sure I understand what deleting a patch proxy resource means. To me, there are two ideas around this:

(1) Deleting an entry in the DB that stores _conceptual_ information around "patch _XYZ_ can be applied to Container App _ABC_"

(2) Deleting an entry in the DB that stores _concrete_ information around "patch _XYZ_ is actively being applied to Container App _ABC_" and halting the process of this new image being patched, pushed and propagated to Container App _ABC_

It seems like it should be the latter specifically because we allow a _Deleting_ provisioning state for the patch proxy resource, but it would be good to get some clarity here.

Reply:
Agree with Cormac's points here. Deleting a patch should possibly be a LRO, depending on the amount of work that is needed. I'm not sure the exact time limit for the request, but IIRC we want to be able to respond within a second. If this cannot be done, then lets consider long running operation / async response here.

And yes, I would also like to understand the difference here. Specifically I am curious why this API seems to indicate that each patch resource (w/ ARM Resource ID of `/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.App/builders/{builderName}/patches/{patchName}`) does not correlate with a potential patch operation. E.g. it seems as though each patch resource can have multiple patch ids in a list under it? Just want to understand why that is, since it is different than the build proxy resources that each correlate to one build

Reply:
Thanks for the feedbacks and questions here! I fogor about the corner case when one patcher is having multiple patchIds related to it. 

The current delete endpoint is not doing any operation on the patch proxy resource (i.e. updating/changing the config). This operation is just removing the specific not-applied patch item from the DB so that user won't see it and can skip it when they run `patch list` next time. 

The main reason for allowing multiple patch id under one patcher is because patching is a long running operation. I think it makes sense if the user want to apply multiple patches to multiple apps in 1 cluster. Therefore, we have 1 proxy resource which may have multiple patches accepted and running/waiting to be applied under one patch resource.

But one of the cases that I forgot was allowing user to remove a "queued" (not started) patch from the patcher. Do you guys think if we need to have this feature?

And yes the arm resource id should have the specific patching execution specified.

Reply:
I personally feel it's too confusing to maintain `patchName` and `patchId` as separate concepts and is equally confusing to separately maintain the concepts of "compute responsible for hosting the patching" and "the process of patching an image".

As @pauld-msft alluded to in his comment, I feel we should keep this as similar as possible to the build proxy resource that he created, where a single build compute is spun up to carry out a singular build. In this case, a single patch compute should be spun up to carry out a singular patch of an image. This allows us to represent each "process of patching an image" with a single "compute responsible for hosting for patching", similar to what we do for the build proxy resource and clears up a lot of confusion about what is meant by `patchName` vs. `patchId` and maintaining these different values.

If we do take this approach, let's say we settle on `patchId` as the unique identifier that represents a single patch that would occur on a single compute; this still allows users to unique identify a single patch they want to apply, and this allows us to spin up the corresponding compute when the user requests for a patch to be applied. I believe this is the same as what we do for the build proxy resource (_i.e._, spinning up the compute when it's needed for a process).

The only caveat of this approach is that it doesn't allow bulk patching within a singular compute, which I don't necessarily think is a bad thing. In this case, rather than having a singular compute responsible for 10 patches, we would have 10 computes each responsible for a single patch, which can be done concurrently compared to the sequential approach with singular compute.

Please let me know if this makes sense and/or if you have any follow-up questions or concerns.